File: runme/mcgen/local/datan/datsrc/avtble.c
#include <math.h>
#include "datsrc.h"

void LIBFUNC avtble_(double const P_T *x, integer const P_T *ni,
             integer const P_T *nj, integer const P_T *nk, double P_T *xb,
             double P_T *xbi, double P_T *xbj, double P_T *xbij,
             double P_T *q, double P_T *s, double P_T *f, integer P_T *ndf,
             double P_T *a)
{
    /* System generated locals */
    integer x_dim1, x_dim2, x_offset, xbij_dim1, xbij_offset, i__1, i__2, 
            i__3;
    double d__1;

    /* Local variables */
    integer i, j, k, l;
    double df[6], ai, ak, aj;

    /* Parameter adjustments */
    --xbi;
    xbij_dim1 = *ni;
    xbij_offset = xbij_dim1 + 1;
    xbij -= xbij_offset;
    --xbj;
    x_dim1 = *ni;
    x_dim2 = *nj;
    x_offset = x_dim1 * (x_dim2 + 1) + 1;
    x -= x_offset;
    --q;
    --s;
    --f;
    --ndf;
    --a;

    /* Function Body */
    ai = (double) (*ni);
    ak = (double) (*nk);
    aj = (double) (*nj);
/* compute means */
    *xb = 0.;
    i__1 = *ni;
    for (i = 1; i <= i__1; ++i) {
        xbi[i] = 0.;
    }
    i__1 = *nj;
    for (j = 1; j <= i__1; ++j) {
        xbj[j] = 0.;
        i__2 = *ni;
        for (i = 1; i <= i__2; ++i) {
            xbij[i + j * xbij_dim1] = 0.;
        }
    }
    i__2 = *ni;
    for (i = 1; i <= i__2; ++i) {
        i__1 = *nj;
        for (j = 1; j <= i__1; ++j) {
            i__3 = *nk;
            for (k = 1; k <= i__3; ++k) {
                xbij[i + j * xbij_dim1] += x[i + (j + k * x_dim2) * x_dim1];
            }
            xbij[i + j * xbij_dim1] /= ak;
        }
    }
    i__1 = *ni;
    for (i = 1; i <= i__1; ++i) {
        i__2 = *nj;
        for (j = 1; j <= i__2; ++j) {
            xbi[i] += xbij[i + j * xbij_dim1];
        }
        xbi[i] /= aj;
    }
    i__1 = *nj;
    for (j = 1; j <= i__1; ++j) {
        i__2 = *ni;
        for (i = 1; i <= i__2; ++i) {
            xbj[j] += xbij[i + j * xbij_dim1];
        }
        xbj[j] /= ai;
        *xb += xbj[j];
    }
    *xb /= aj;
/* compute sums of squares */
    for (l = 1; l <= 6; ++l) {
        q[l] = 0.;
    }
    i__1 = *ni;
    for (i = 1; i <= i__1; ++i) {
/* Computing 2nd power */
        d__1 = xbi[i] - *xb;
        q[1] += d__1 * d__1;
    }
    q[1] = aj * ak * q[1];
    i__1 = *nj;
    for (j = 1; j <= i__1; ++j) {
/* Computing 2nd power */
        d__1 = xbj[j] - *xb;
        q[2] += d__1 * d__1;
    }
    q[2] = ai * ak * q[2];
    i__1 = *ni;
    for (i = 1; i <= i__1; ++i) {
        i__2 = *nj;
        for (j = 1; j <= i__2; ++j) {
/* Computing 2nd power */
            d__1 = xbij[i + j * xbij_dim1] + *xb - xbi[i] - xbj[j];
            q[3] += d__1 * d__1;
            i__3 = *nk;
            for (k = 1; k <= i__3; ++k) {
/* Computing 2nd power */
                d__1 = x[i + (j + k * x_dim2) * x_dim1] -
                       xbij[i + j * xbij_dim1];
                q[5] += d__1 * d__1;
/* Computing 2nd power */
                d__1 = x[i + (j + k * x_dim2) * x_dim1] - *xb;
                q[6] += d__1 * d__1;
            }
        }
    }
    q[3] = ak * q[3];
    q[4] = q[2] + q[3];
/* determine degrees of freedom */
    df[0] = ai - 1.;
    df[1] = aj - 1.;
    df[2] = (ai - 1.) * (aj - 1.);
    df[3] = df[1] + df[2];
    df[4] = ai * aj * (ak - 1.);
    df[5] = ai * aj * ak - 1.;
/* mean squares */
    for (l = 1; l <= 6; ++l) {
        s[l] = q[l] / df[l - 1];
    }
/* F quotients */
    for (l = 1; l <= 4; ++l) {
        f[l] = s[l] / s[5];
    }
/* degrees of freedom as integers */
    for (l = 1; l <= 6; ++l) {
        ndf[l] = nint(df[l - 1]);
    }
/* levels of significance */
    for (l = 1; l <= 4; ++l) {
        a[l] = 1. - scftst_(&f[l], &ndf[l], &ndf[5]);
    }
    return;
} /* avtble_ */

File: runme/mcgen/local/datan/datsrc/avtble.f
      SUBROUTINE AVTBLE(X,NI,NJ,NK,XB,XBI,XBJ,XBIJ,Q,S,F,NDF,A)
      IMPLICIT DOUBLE PRECISION (A-H,O-Z)
      DIMENSION X(NI,NJ,NK),XBI(NI),XBJ(NJ),XBIJ(NI,NJ),Q(6),DF(6)
      DIMENSION S(6),F(4),NDF(6),A(4)
      PARAMETER(ZERO=0.D0, ONE=1.D0)
      AI=DBLE(NI)
      AK=DBLE(NK)
      AJ=DBLE(NJ)
C compute means
      XB=ZERO
      DO 10 I=1,NI
        XBI(I)=ZERO
   10 CONTINUE
      DO 20 J=1,NJ
        XBJ(J)=ZERO
        DO 20 I=1,NI
          XBIJ(I,J)=ZERO
   20 CONTINUE
      DO 40 I=1,NI
        DO 40 J=1,NJ
          DO 30 K=1,NK
            XBIJ(I,J) =XBIJ(I,J)+X(I,J,K)
   30     CONTINUE
          XBIJ(I,J) =XBIJ(I,J)/AK
   40 CONTINUE
      DO 60 I=1,NI
        DO 50 J=1,NJ
          XBI(I)=XBI(I)+XBIJ(I,J)
   50   CONTINUE
        XBI(I)=XBI(I)/AJ
   60 CONTINUE
      DO 80 J=1,NJ
        DO 70 I=1,NI
          XBJ(J)=XBJ(J)+XBIJ(I,J)
   70   CONTINUE
        XBJ(J)=XBJ(J)/AI
        XB=XB+XBJ(J)
   80 CONTINUE
      XB=XB/AJ
C compute sums of squares
      DO 90 L=1,6
        Q(L)=ZERO
   90 CONTINUE
      DO 100 I=1,NI
        Q(1)=Q(1)+(XBI(I)-XB)**2
  100 CONTINUE
      Q(1)=AJ*AK*Q(1)
      DO 110 J=1,NJ
        Q(2)=Q(2)+(XBJ(J)-XB)**2
  110 CONTINUE
      Q(2)=AI*AK*Q(2)
      DO 120 I=1,NI
        DO 120 J=1,NJ
          Q(3)=Q(3)+(XBIJ(I,J)+XB-XBI(I)-XBJ(J))**2
          DO 120 K=1,NK
            Q(5)=Q(5)+(X(I,J,K)-XBIJ(I,J))**2
            Q(6)=Q(6)+(X(I,J,K)-XB)**2
  120 CONTINUE
      Q(3) =AK*Q(3)
      Q(4)=Q(2)+Q(3)
C determine degrees of freedom
      DF(1)=AI-ONE
      DF(2)=AJ-ONE
      DF(3)=(AI-ONE)*(AJ-ONE)
      DF(4)=DF(2)+DF(3)
      DF(5)=AI*AJ*(AK-ONE)
      DF(6)=AI*AJ*AK-ONE
C mean squares
      DO 130 L=1,6
        S(L)=Q(L)/DF(L)
  130 CONTINUE
C F quotients
      DO 140 L=1,4
        F(L)=S(L)/S(5)
  140 CONTINUE
C degrees of freedom as integers
      DO 150 L=1,6
        NDF(L)=NINT(DF(L))
  150 CONTINUE
C levels of significance
      DO 160 L=1,4
        A(L)=ONE-SCFTST(F(L),NDF(L),NDF(5))
  160 CONTINUE
      END
